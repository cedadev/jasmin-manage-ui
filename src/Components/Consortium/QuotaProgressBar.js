import React, { useEffect, useRef, useState } from 'react';

import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Overlay from 'react-bootstrap/Overlay';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Row from 'react-bootstrap/Row';
import Table from 'react-bootstrap/Table';
import Tooltip from 'react-bootstrap/Tooltip';

import classNames from 'classnames';

import { Status } from '../../rest-resource';

import { useResources } from '../../api';

import { SpinnerWithText, formatAmount, sortByKey } from '../utils';

import '../../css/quota-progress-bar.css';


const QuotaItemProgressBar = ({ status, now, units, ...props }) => {
    const [showTooltip, setShowTooltip] = useState(false);
    const target = useRef(null);
    return (<>
        <ProgressBar
            {...props}
            ref={target}
            now={now}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
        />
        <Overlay target={target.current} show={showTooltip} placement="bottom">
            {props => (
                <Tooltip {...props}>
                    {status}: {formatAmount(now, units)}
                </Tooltip>
            )}
        </Overlay>
    </>);
};


const QuotaAmountIndicator = ({ max, now, units }) => {
    const [showTooltip, setShowTooltip] = useState(false);
    const target = useRef(null);
    return (<>
        <div
            ref={target}
            className="indicator-line"
            style={{ left: `${(now / max) * 100}%` }}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
        />
        <Overlay target={target.current} show={showTooltip} placement="bottom">
            {props => (
                <Tooltip {...props}>
                    Quota amount: {formatAmount(now, units)}
                </Tooltip>
            )}
        </Overlay>
    </>);
};


export const QuotaProgressBar = ({
    quota,
    resource,
    // If given, this requirement is shown alongside the quota amounts
    requirement,
    CaptionComponent
}) => {
    const units = resource.data.units;
    const quotaAmount = quota.data.amount;

    // Start with the totals from the quota
    let {
        total_approved: totalApproved,
        total_awaiting_provisioning: totalAwaiting,
        total_provisioned: totalProvisioned
    } = quota.data;
    let requirementAmount = 0;

    // Correct the amounts if we were given a requirement
    if( requirement ) {
        requirementAmount = requirement.data.amount;
        switch( requirement.data.status ) {
            case "APPROVED":
                totalApproved -= requirementAmount;
                break;
            case "AWAITING_PROVISIONING":
                totalAwaiting -= requirementAmount;
                break;
            case "PROVISIONED":
                totalProvisioned -= requirementAmount;
                break;
        }
    }

    // Work out the total commitments
    const totalCommitted = totalApproved + totalAwaiting + totalProvisioned;
    const totalWithRequirement = totalCommitted + requirementAmount;
    const committedPercentage = Math.round((totalCommitted / quotaAmount) * 100);
    // Calculate the remaining amount after the requirement is approved
    const remainingAmount = quotaAmount - totalWithRequirement;
    // Calculate the max for the progress bar
    const progressBarMax = Math.max(quotaAmount, totalWithRequirement);

    return (
        <div className="quota-progress-bar-container">
            <CaptionComponent className="text-center">
                {formatAmount(totalCommitted, units)}{" "}
                of{" "}
                {formatAmount(quotaAmount, units)}{" "}
                committed ({committedPercentage}%)
            </CaptionComponent>
            <div className="quota-progress-bar">
                <ProgressBar>
                    <QuotaItemProgressBar
                        status="Provisioned"
                        variant="info"
                        max={progressBarMax}
                        now={totalProvisioned}
                        units={units}
                    />
                    <QuotaItemProgressBar
                        status="Awaiting Provisioning"
                        variant="warning"
                        max={progressBarMax}
                        now={totalAwaiting}
                        units={units}
                    />
                    <QuotaItemProgressBar
                        status="Approved"
                        variant="success"
                        max={progressBarMax}
                        now={totalApproved}
                        units={units}
                    />
                    {requirementAmount > 0 && (
                        <QuotaItemProgressBar
                            status="This requirement"
                            variant="secondary"
                            max={progressBarMax}
                            now={requirementAmount}
                            units={units}
                        />
                    )}
                    {/* If there is still headroom in the quota, but in a bar for it */}
                    {remainingAmount > 0 && (
                        <QuotaItemProgressBar
                            status="Remaining"
                            variant="transparent"
                            max={progressBarMax}
                            now={remainingAmount}
                            units={units}
                        />
                    )}
                </ProgressBar>
                {/* If the progress bar goes beyond the quota, show a marker at the quota */}
                {progressBarMax > quotaAmount && (
                    <QuotaAmountIndicator
                        max={progressBarMax}
                        now={quotaAmount}
                        units={units}
                    />
                )}
            </div>
        </div>
    );
};
