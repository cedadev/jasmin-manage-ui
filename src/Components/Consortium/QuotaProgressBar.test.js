// QuotaProgressBar.test.js
import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { QuotaProgressBar } from "./QuotaProgressBar";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";

// Mock formatAmount utility
jest.mock("../utils", () => ({
  formatAmount: (value, units) => `${value}${units}`,
}));

describe("QuotaProgressBar Component", () => {
  const mockQuota = {
    data: {
      amount: 100,
      total_approved: 30,
      total_awaiting_provisioning: 20,
      total_provisioned: 10,
    },
  };

  const mockResource = {
    data: {
      units: "GB",
    },
  };

  const defaultProps = {
    quota: mockQuota,
    resource: mockResource,
    CaptionComponent: "div",
  };

  it("Should match the snapshot", () => {
    const component = renderer.create(<QuotaProgressBar {...defaultProps} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should render basic progress bar without requirement", () => {
    render(<QuotaProgressBar {...defaultProps} />);

    // Check caption shows correct total
    expect(screen.getByText(/60GB of 100GB committed/)).toBeInTheDocument();
    expect(screen.getByText(/\(60%\)/)).toBeInTheDocument();

    // Verify all progress bar sections are present with correct values
    const progressBars = screen.getAllByRole("progressbar");
    expect(progressBars).toHaveLength(4); // 3 status bars + 1 remaining

    // Verify individual progress bar sections
    const [provisioned, awaiting, approved, remaining] = progressBars;

    // Check provisioned section
    expect(provisioned).toHaveClass("bg-info");
    expect(provisioned).toHaveAttribute("aria-valuenow", "10");

    // Check awaiting section
    expect(awaiting).toHaveClass("bg-warning");
    expect(awaiting).toHaveAttribute("aria-valuenow", "20");

    // Check approved section
    expect(approved).toHaveClass("bg-success");
    expect(approved).toHaveAttribute("aria-valuenow", "30");

    // Check remaining section
    expect(remaining).toHaveClass("bg-transparent");
    expect(remaining).toHaveAttribute("aria-valuenow", "40");
  });

  it("Should handle APPROVED requirement correctly", () => {
    const requirement = {
      data: {
        amount: 20,
        status: "APPROVED",
      },
    };

    render(<QuotaProgressBar {...defaultProps} requirement={requirement} />);

    // Total committed should exclude requirement amount from approved
    expect(screen.getByText(/40GB of 100GB committed/)).toBeInTheDocument();
  });

  it("Should handle AWAITING_PROVISIONING requirement correctly", () => {
    const requirement = {
      data: {
        amount: 10,
        status: "AWAITING_PROVISIONING",
      },
    };

    render(<QuotaProgressBar {...defaultProps} requirement={requirement} />);

    // Total committed should exclude requirement amount from awaiting
    expect(screen.getByText(/50GB of 100GB committed/)).toBeInTheDocument();
  });

  it("Should handle PROVISIONED requirement correctly", () => {
    const requirement = {
      data: {
        amount: 5,
        status: "PROVISIONED",
      },
    };

    render(<QuotaProgressBar {...defaultProps} requirement={requirement} />);

    // Total committed should exclude requirement amount from provisioned
    expect(screen.getByText(/55GB of 100GB committed/)).toBeInTheDocument();
  });

  it("Should show quota indicator when total exceeds quota", () => {
    const largeQuota = {
      data: {
        amount: 50,
        total_approved: 60,
        total_awaiting_provisioning: 0,
        total_provisioned: 0,
      },
    };

    // Destructure container from render
    const { container } = render(
      <QuotaProgressBar {...defaultProps} quota={largeQuota} />
    );

    // Find quota-progress-bar div then find indicator line within it
    const progressBarDiv = container.querySelector(".quota-progress-bar");
    const indicatorLine = progressBarDiv.querySelector(".indicator-line");

    // Verify indicator exists and position
    expect(indicatorLine).toBeInTheDocument();
    expect(indicatorLine).toHaveStyle({
      left: "83.33333333333334%", // (50/60 * 100%)
    });
  });

  it("Should show tooltips on hover", async () => {
    const user = userEvent.setup();
    render(<QuotaProgressBar {...defaultProps} />);

    // Get all progress bars and find the provisioned one (first with bg-info class)
    const progressBars = screen.getAllByRole("progressbar");
    const provisionedBar = progressBars.find((bar) =>
      bar.classList.contains("bg-info")
    );

    // Hover over provisioned section
    await act(async () => {
      await user.hover(provisionedBar);
    });

    // Wait for tooltip
    await waitFor(() => {
      expect(screen.getByText(/Provisioned: 10GB/)).toBeInTheDocument();
    });

    // Unhover
    await act(async () => {
      await user.unhover(provisionedBar);
    });

    // Tooltip should disappear
    await waitFor(() => {
      expect(screen.queryByText(/Provisioned: 10GB/)).not.toBeInTheDocument();
    });
  });

  it("Should show remaining amount when less than quota", () => {
    const smallQuota = {
      data: {
        amount: 100,
        total_approved: 20,
        total_awaiting_provisioning: 10,
        total_provisioned: 10,
      },
    };

    render(<QuotaProgressBar {...defaultProps} quota={smallQuota} />);

    // Should show remaining bar
    const progressBars = screen.getAllByRole("progressbar");
    expect(progressBars).toHaveLength(4); // 3 status + 1 remaining
  });

  it("Should show and hide tooltip on indicator line hover", async () => {
    const largeQuota = {
      data: {
        amount: 50,
        total_approved: 60,
        total_awaiting_provisioning: 0,
        total_provisioned: 0,
      },
    };

    const { container } = render(
      <QuotaProgressBar {...defaultProps} quota={largeQuota} />
    );

    // Find indicator line
    const progressBarDiv = container.querySelector(".quota-progress-bar");
    const indicatorLine = progressBarDiv.querySelector(".indicator-line");
    expect(indicatorLine).toBeInTheDocument();

    const user = userEvent.setup();

    // Mouse enter - show tooltip
    await act(async () => {
      await user.hover(indicatorLine);
    });
    expect(screen.getByText(/Quota amount: 50GB/)).toBeInTheDocument();

    // Mouse leave - hide tooltip
    await act(async () => {
      await user.unhover(indicatorLine);
    });
    await waitFor(() => {
      expect(screen.queryByText(/Quota amount: 50GB/)).not.toBeInTheDocument();
    });
  });
});
