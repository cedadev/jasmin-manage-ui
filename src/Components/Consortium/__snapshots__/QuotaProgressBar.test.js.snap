// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuotaProgressBar Component Should match the snapshot 1`] = `
<div
  className="quota-progress-bar-container"
>
  <div
    className="text-center"
  >
    60GB
     
    of
     
    100GB
     
    committed (
    60
    %)
  </div>
  <div
    className="quota-progress-bar"
  >
    <div
      className="progress"
    >
      <div
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={10}
        className="progress-bar bg-info"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="progressbar"
        style={
          {
            "width": "10%",
          }
        }
      />
      <div
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={20}
        className="progress-bar bg-warning"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="progressbar"
        style={
          {
            "width": "20%",
          }
        }
      />
      <div
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={30}
        className="progress-bar bg-success"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="progressbar"
        style={
          {
            "width": "30%",
          }
        }
      />
      <div
        aria-valuemax={100}
        aria-valuemin={0}
        aria-valuenow={40}
        className="progress-bar bg-transparent"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="progressbar"
        style={
          {
            "width": "40%",
          }
        }
      />
    </div>
  </div>
</div>
`;
